ostd/src/sync/spin.rs

```diff
@@ -16,65 +17,87 @@ use crate::{
 };
 
 /// A spin lock.
-pub struct SpinLock<T: ?Sized> {
+#[repr(transparent)]
+pub struct SpinLock<T: ?Sized, G = PreemptDisabled> {
+    phantom: PhantomData<G>,
+    /// Only the last field of a struct may have a dynamically sized type
+    /// That's why SpinLockInner is put in the last field
+    inner: SpinLockInner<T>,
+}
+
+struct SpinLockInner<T: ?Sized> {
     lock: AtomicBool,
     val: UnsafeCell<T>,
 }
 
-impl<T> SpinLock<T> {
+pub trait Guardian {
+    type Guard;
+
+    /// Return a guard object
+    fn guard() -> Self::Guard;
+}
+
+/// for lock disabling preempt
+pub struct PreemptDisabled;
+
+impl Guardian for PreemptDisabled {
+    type Guard = DisablePreemptGuard;
+
+    fn guard() -> Self::Guard {
+        disable_preempt()
+    }
+}
+
+/// for lock disabling Irq
+pub struct LocalIrqDisabled;
+
+impl Guardian for LocalIrqDisabled {
+    type Guard = DisabledLocalIrqGuard;
+
+    fn guard() -> Self::Guard {
+        disable_local()
+    }
+}
+
+impl<T, G: Guardian> SpinLock<T, G> {
     /// Creates a new spin lock.
     pub const fn new(val: T) -> Self {
-        Self {
+        let lock_inner = SpinLockInner {
             lock: AtomicBool::new(false),
             val: UnsafeCell::new(val),
+        };
+        Self {
+            phantom: PhantomData,
+            inner: lock_inner,
         }
     }
 }
 
-impl<T: ?Sized> SpinLock<T> {
-    /// Acquires the spin lock with disabling the local IRQs. This is the most secure
-    /// locking way.
-    ///
-    /// This method runs in a busy loop until the lock can be acquired.
-    /// After acquiring the spin lock, all interrupts are disabled.
-    pub fn lock_irq_disabled(&self) -> SpinLockGuard<T> {
-        let guard = disable_local();
-        self.acquire_lock();
-        SpinLockGuard_ {
-            lock: self,
-            inner_guard: InnerGuard::IrqGuard(guard),
-        }
-    }
-
-    /// Tries acquiring the spin lock immedidately with disabling the local IRQs.
-    pub fn try_lock_irq_disabled(&self) -> Option<SpinLockGuard<T>> {
-        let irq_guard = disable_local();
-        if self.try_acquire_lock() {
-            let lock_guard = SpinLockGuard_ {
-                lock: self,
-                inner_guard: InnerGuard::IrqGuard(irq_guard),
-            };
-            return Some(lock_guard);
-        }
-        None
+impl<T: ?Sized> SpinLock<T, PreemptDisabled> {
+    /// convert spinlock from PreemptDisabled to LocalIrqDisabled
+    pub fn disable_irq(&self) -> &SpinLock<T, LocalIrqDisabled> {
+        // SAFETY: The memory layout of the source and target types, which only differ in the `phantom` field, are identical because
+        // (1) `PhantomData` is zero-sized, (2) the struct layout follows `repr(C)`.
+        unsafe { core::mem::transmute(self) }
```

Comment:
It seems that the Rust documentation [suggests](https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives) using pointer reborrows instead of `transmute` in this case.

Bssides, the documentation needs to be updated as well. Moreover, the long lines need to be broken into multiple lines.

```rust
// SAFETY:
// 1. The types `SpinLock<T, PreemptDisabled>`, `SpinLockInner<T>` and `SpinLock<T,
//    IrqDisabled>` have the same memory layout guaranteed by `#[repr(transparent)]`.
// 2. The specified memory location can be borrowed as an immutable reference for the
//    specified lifetime.
unsafe { &*ptr::from_ref(self).cast() }
```

Reply:
Thanks for your instructions! The bug above is fixed under your guidance, and it could pass all tests now.

I would go through all documentations and modify carefully, that would probably not cause new problems. For this comment, though, I change from `core::mem::transmute(self)` to `&*ptr::from_ref(self).cast()`, then compilation fails:
![屏幕截图 2024-08-14 101950](https://github.com/user-attachments/assets/1df909fc-f3ec-40ea-a025-5eb86925d31c)
It seems that `case` could not be used by variables with `T: ?Sized`. In this situation, could I still use `transmute`, or are there methods fixing this error for `case`?

Reply:
Oh, I think here `ptr::cast` doesn't work because it needs to make sure that the two pointers have the same [pointer metadata](https://doc.rust-lang.org/std/ptr/trait.Pointee.html#pointer-metadata). So it can only handle sized pointers, because in this case the pointer metadata must be `()`.

We can just use `as` expressions in this case, so the compiler will check if the two pointers have the same metadata. We should be able to just write:
```rust
    let ptr = self as *const SpinLock<T, PreemptDisabled>;
    let ptr = ptr as *const SpinLock<T, IrqDisabled>;

    // SAFETY: ...
    unsafe { &*ptr }
```